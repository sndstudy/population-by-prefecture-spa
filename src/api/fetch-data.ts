import axios from 'axios';

const RESAS_API_ENDPOINT = 'https://opendata.resas-portal.go.jp/';
const API_KEY = process.env['REACT_APP_RESAS_API_KEY'] || '';

const instance = axios.create({
  baseURL: RESAS_API_ENDPOINT,
  timeout: 1000,
  headers: { 'X-API-KEY': API_KEY },
});

export type PrefecturesResponse = {
  message: string;
  result: Prefectures[];
};

export type Prefectures = {
  prefCode: string;
  prefName: string;
};

export type PopulationResponse = {
  message: string;
  result: Population;
};

export type Population = {
  boundaryYear: number;
  data: {
    label: string;
    data: {
      year: number;
      value: number;
      rate?: number;
    }[];
  }[];
};

export const fetchPrefectures = async () => {
  const { data } = await instance.get<
    PrefecturesResponse | { statusCode: string }
  >('api/v1/prefectures');

  // エラーの場合、ステータスコードがレスポンスボディに設定されているため「statusCode」の有無でエラーを判定する
  // https://opendata.resas-portal.go.jp/docs/api/v1/detail/index.html
  if ('statusCode' in data) {
    return Promise.reject();
  }

  return data;
};

export const fetchPopulationData = async ({
  queryKey,
}: {
  queryKey: string[];
}) => {
  const [, prefCode] = queryKey;

  const { data } = await instance.get<
    PopulationResponse | { statusCode: string }
  >('api/v1/population/composition/perYear', {
    params: {
      prefCode,
    },
  });

  // エラーの場合、ステータスコードがレスポンスボディに設定されているため「statusCode」の有無でエラーを判定する
  // https://opendata.resas-portal.go.jp/docs/api/v1/detail/index.html
  if ('statusCode' in data) {
    return Promise.reject();
  }

  return data;
};
