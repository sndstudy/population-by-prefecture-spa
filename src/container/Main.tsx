import { ChangeEvent, useState } from 'react';
import { useQuery } from 'react-query';
import { fetchPrefectures } from '../api/fetch-data';
import CheckBox from '../presentational/CheckBox';
import styles from '../styles/Main.module.scss';

const Main = () => {
  const { data, isLoading } = useQuery('fetchPrefectures', fetchPrefectures);
  const [selectedPrefectures, setSelectedPrefectures] = useState<string[]>([]);

  const change = (e: ChangeEvent<HTMLInputElement>) => {
    if (e.target.checked) {
      setSelectedPrefectures((prev) => [...prev, e.target.value]);
    } else {
      setSelectedPrefectures((prev) =>
        prev.filter((v) => v !== e.target.value),
      );
    }
  };

  if (isLoading) {
    return <>データ取得中</>;
  }

  return (
    <>
      <div className={styles.title}>都道府県別 総人口</div>
      <div className={styles.wrapper}>
        <div className={styles.prefectures_area}>
          <div className={styles.section_title}>都道府県</div>
          {data &&
            data.result.map((prefecture) => {
              return (
                <CheckBox
                  key={prefecture.prefCode}
                  value={prefecture.prefCode}
                  label={prefecture.prefName}
                  checked={selectedPrefectures.some(
                    (selectedPrefecture) =>
                      selectedPrefecture === String(prefecture.prefCode),
                  )}
                  onChange={change}
                />
              );
            })}
        </div>
      </div>
    </>
  );
};

export default Main;
