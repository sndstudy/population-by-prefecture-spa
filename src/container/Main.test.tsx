import { render, screen, waitFor } from '@testing-library/react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import Main from './Main';
import { PrefecturesResponse } from '../api/fetch-data';
import { QueryClient, QueryClientProvider } from 'react-query';
import userEvent from '@testing-library/user-event';

const populationData = [
  {
    prefCode: '1',
    data: {
      message: 'テストメッセージ',
      result: {
        boundaryYear: 2020,
        data: [
          {
            label: '総人口',
            data: [
              {
                year: 1980,
                value: 1000,
              },
              {
                year: 1985,
                value: 2000,
              },
              {
                year: 1990,
                value: 3000,
              },
            ],
          },
        ],
      },
    },
  },
  {
    prefCode: '2',
    data: {
      message: 'テストメッセージ',
      result: {
        boundaryYear: 2020,
        data: [
          {
            label: '総人口',
            data: [
              {
                year: 1980,
                value: 4000,
              },
              {
                year: 1985,
                value: 5000,
              },
              {
                year: 1990,
                value: 6000,
              },
            ],
          },
        ],
      },
    },
  },
  {
    prefCode: '3',
    data: {
      message: null,
      result: {
        boundaryYear: 2020,
        data: [
          {
            label: '総人口',
            data: [
              {
                year: 1980,
                value: 7000,
              },
              {
                year: 1985,
                value: 8000,
              },
              {
                year: 1990,
                value: 9000,
              },
            ],
          },
        ],
      },
    },
  },
];

const server = setupServer(
  rest.get(
    'https://opendata.resas-portal.go.jp/api/v1/prefectures',
    (req, res, ctx) => {
      return res(
        ctx.json<PrefecturesResponse>({
          message: 'テストメッセージ',
          result: [
            { prefCode: '1', prefName: '北海道' },
            { prefCode: '2', prefName: '青森県' },
            { prefCode: '3', prefName: '岩手県' },
          ],
        }),
      );
    },
  ),
  rest.get(
    'https://opendata.resas-portal.go.jp/api/v1/population/composition/perYear',
    (req, res, ctx) => {
      const prefCode = req.url.searchParams.get('prefCode');
      const resData = populationData.find((v) => v.prefCode === prefCode);
      return res(ctx.json(resData?.data));
    },
  ),
);

describe('Mainのテスト', () => {
  beforeAll(() => {
    server.listen();
  });
  afterAll(() => {
    server.close();
  });

  const queryClient = new QueryClient();

  it('初期描画', async () => {
    const { asFragment } = render(
      <QueryClientProvider client={queryClient}>
        <Main />
      </QueryClientProvider>,
    );

    await waitFor(() => {
      expect(screen.getByText('データ取得中')).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();

    await waitFor(() => {
      const checkBoxList = screen.getAllByRole('checkbox');
      expect(checkBoxList.length).toBe(3);
      checkBoxList.forEach((v) => {
        // チェックが付いていないことを確認
        expect(v).not.toBeChecked();
      });
    });

    expect(screen.queryByText('データ取得中')).toBeNull();
    expect(asFragment()).toMatchSnapshot();
  });

  it('都道府県を選択(グラフ描画)', async () => {
    const { asFragment } = render(
      <QueryClientProvider client={queryClient}>
        <Main />
      </QueryClientProvider>,
    );

    await waitFor(() => {
      const checkBoxList = screen.getAllByRole('checkbox');
      expect(checkBoxList.length).toBe(3);
      checkBoxList.forEach((v) => {
        // チェックが付いていないことを確認
        expect(v).not.toBeChecked();
      });
    });

    expect(screen.queryByText('データ取得中')).toBeNull();
    expect(asFragment()).toMatchSnapshot();

    const checkbox = screen.getByLabelText('北海道');

    userEvent.click(checkbox);
    expect(checkbox).toBeChecked();

    await waitFor(() => {
      // チェックボックスとグラフのラベル
      expect(screen.getAllByText('北海道').length).toEqual(2);
    });

    // グラフが描画されていることを確認
    expect(asFragment()).toMatchSnapshot();
  });
});
