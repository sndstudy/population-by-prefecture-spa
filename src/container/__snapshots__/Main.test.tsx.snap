// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mainのテスト 初期描画 1`] = `
<DocumentFragment>
  <div
    class="title"
  >
    都道府県別 総人口
  </div>
  <div
    class="wrapper"
  >
    <div
      class="prefectures_area"
    >
      <div
        class="section_title"
      >
        都道府県選択
      </div>
      データ取得中
    </div>
    <div
      class="graph_area"
    >
      <div
        data-highcharts-chart="0"
        style="height: 100%; overflow: hidden;"
      >
        <div
          class="highcharts-container "
          dir="ltr"
          id="highcharts-0"
          style="position: relative; overflow: hidden; width: 600px; height: 400px; text-align: left; line-height: normal; z-index: 0; user-select: none; outline: none;"
        >
          <svg
            aria-label="都道府県別 総人口数"
            class="highcharts-root"
            height="400"
            role="img"
            style="font-family: \\"Lucida Grande\\", \\"Lucida Sans Unicode\\", Arial, Helvetica, sans-serif; font-size: 12px;"
            version="1.1"
            viewBox="0 0 600 400"
            width="600"
            xmlns="http://www.w3.org/2000/svg"
          >
            <desc>
              Created with Highcharts 10.3.3
            </desc>
            <defs>
              <clippath
                id="highcharts-1-"
              >
                <rect
                  fill="none"
                  height="250"
                  width="580"
                  x="0"
                  y="0"
                />
              </clippath>
            </defs>
            <rect
              class="highcharts-background"
              fill="#ffffff"
              height="400"
              rx="0"
              ry="0"
              width="600"
              x="0"
              y="0"
            />
            <rect
              class="highcharts-plot-background"
              fill="none"
              height="250"
              width="580"
              x="10"
              y="50"
            />
            <rect
              class="highcharts-plot-border"
              data-z-index="1"
              fill="none"
              height="250"
              stroke="#cccccc"
              stroke-width="0"
              width="580"
              x="10"
              y="50"
            />
            <g
              class="highcharts-series-group"
              data-z-index="3"
            />
            <text
              class="highcharts-title"
              data-z-index="4"
              style="color: rgb(51, 51, 51); font-size: 18px; fill: #333333;"
              text-anchor="middle"
              x="300"
              y="64"
            >
              都道府県別 総人口数
            </text>
            <text
              class="highcharts-subtitle"
              data-z-index="4"
              style="color: rgb(102, 102, 102); fill: #666666;"
              text-anchor="middle"
              x="300"
              y="50"
            />
            <text
              class="highcharts-caption"
              data-z-index="4"
              style="color: rgb(102, 102, 102); fill: #666666;"
              text-anchor="start"
              x="10"
              y="312"
            />
            <g
              class="highcharts-legend highcharts-no-tooltip"
              data-z-index="7"
              visibility="hidden"
            >
              <rect
                class="highcharts-legend-box"
                fill="none"
                height="8"
                rx="0"
                ry="0"
                stroke="#999999"
                stroke-width="0"
                width="8"
                x="0"
                y="0"
              />
              <g
                data-z-index="1"
              >
                <g />
              </g>
            </g>
            <text
              class="highcharts-credits"
              data-z-index="8"
              style="cursor: pointer; color: rgb(153, 153, 153); font-size: 9px; fill: #999999;"
              text-anchor="end"
              x="590"
              y="395"
            >
              Highcharts.com
            </text>
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Mainのテスト 初期描画 2`] = `
<DocumentFragment>
  <div
    class="title"
  >
    都道府県別 総人口
  </div>
  <div
    class="wrapper"
  >
    <div
      class="prefectures_area"
    >
      <div
        class="section_title"
      >
        都道府県選択
      </div>
      <div>
        <label>
          <input
            type="checkbox"
            value="1"
          />
          北海道
        </label>
      </div>
      <div>
        <label>
          <input
            type="checkbox"
            value="2"
          />
          青森県
        </label>
      </div>
      <div>
        <label>
          <input
            type="checkbox"
            value="3"
          />
          岩手県
        </label>
      </div>
    </div>
    <div
      class="graph_area"
    >
      <div
        data-highcharts-chart="0"
        style="height: 100%; overflow: hidden;"
      >
        <div
          class="highcharts-container "
          dir="ltr"
          id="highcharts-0"
          style="position: relative; overflow: hidden; width: 600px; height: 400px; text-align: left; line-height: normal; z-index: 0; user-select: none; outline: none;"
        >
          <svg
            aria-label="都道府県別 総人口数"
            class="highcharts-root"
            height="400"
            role="img"
            style="font-family: \\"Lucida Grande\\", \\"Lucida Sans Unicode\\", Arial, Helvetica, sans-serif; font-size: 12px;"
            version="1.1"
            viewBox="0 0 600 400"
            width="600"
            xmlns="http://www.w3.org/2000/svg"
          >
            <desc>
              Created with Highcharts 10.3.3
            </desc>
            <defs>
              <clippath
                id="highcharts-1-"
              >
                <rect
                  fill="none"
                  height="250"
                  width="580"
                  x="0"
                  y="0"
                />
              </clippath>
            </defs>
            <rect
              class="highcharts-background"
              fill="#ffffff"
              height="400"
              rx="0"
              ry="0"
              width="600"
              x="0"
              y="0"
            />
            <rect
              class="highcharts-plot-background"
              fill="none"
              height="250"
              width="580"
              x="10"
              y="50"
            />
            <rect
              class="highcharts-plot-border"
              data-z-index="1"
              fill="none"
              height="250"
              stroke="#cccccc"
              stroke-width="0"
              width="580"
              x="10"
              y="50"
            />
            <g
              class="highcharts-series-group"
              data-z-index="3"
            />
            <text
              class="highcharts-title"
              data-z-index="4"
              style="color: rgb(51, 51, 51); font-size: 18px; fill: #333333;"
              text-anchor="middle"
              x="300"
              y="64"
            >
              都道府県別 総人口数
            </text>
            <text
              class="highcharts-subtitle"
              data-z-index="4"
              style="color: rgb(102, 102, 102); fill: #666666;"
              text-anchor="middle"
              x="300"
              y="50"
            />
            <text
              class="highcharts-caption"
              data-z-index="4"
              style="color: rgb(102, 102, 102); fill: #666666;"
              text-anchor="start"
              x="10"
              y="312"
            />
            <g
              class="highcharts-legend highcharts-no-tooltip"
              data-z-index="7"
              visibility="hidden"
            >
              <rect
                class="highcharts-legend-box"
                fill="none"
                height="8"
                rx="0"
                ry="0"
                stroke="#999999"
                stroke-width="0"
                width="8"
                x="0"
                y="0"
              />
              <g
                data-z-index="1"
              >
                <g />
              </g>
            </g>
            <text
              class="highcharts-credits"
              data-z-index="8"
              style="cursor: pointer; color: rgb(153, 153, 153); font-size: 9px; fill: #999999;"
              text-anchor="end"
              x="590"
              y="395"
            >
              Highcharts.com
            </text>
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Mainのテスト 都道府県を選択(グラフ描画) 1`] = `
<DocumentFragment>
  <div
    class="title"
  >
    都道府県別 総人口
  </div>
  <div
    class="wrapper"
  >
    <div
      class="prefectures_area"
    >
      <div
        class="section_title"
      >
        都道府県選択
      </div>
      <div>
        <label>
          <input
            type="checkbox"
            value="1"
          />
          北海道
        </label>
      </div>
      <div>
        <label>
          <input
            type="checkbox"
            value="2"
          />
          青森県
        </label>
      </div>
      <div>
        <label>
          <input
            type="checkbox"
            value="3"
          />
          岩手県
        </label>
      </div>
    </div>
    <div
      class="graph_area"
    >
      <div
        data-highcharts-chart="1"
        style="height: 100%; overflow: hidden;"
      >
        <div
          class="highcharts-container "
          dir="ltr"
          id="highcharts-2"
          style="position: relative; overflow: hidden; width: 600px; height: 400px; text-align: left; line-height: normal; z-index: 0; user-select: none; outline: none;"
        >
          <svg
            aria-label="都道府県別 総人口数"
            class="highcharts-root"
            height="400"
            role="img"
            style="font-family: \\"Lucida Grande\\", \\"Lucida Sans Unicode\\", Arial, Helvetica, sans-serif; font-size: 12px;"
            version="1.1"
            viewBox="0 0 600 400"
            width="600"
            xmlns="http://www.w3.org/2000/svg"
          >
            <desc>
              Created with Highcharts 10.3.3
            </desc>
            <defs>
              <clippath
                id="highcharts-3-"
              >
                <rect
                  fill="none"
                  height="250"
                  width="580"
                  x="0"
                  y="0"
                />
              </clippath>
            </defs>
            <rect
              class="highcharts-background"
              fill="#ffffff"
              height="400"
              rx="0"
              ry="0"
              width="600"
              x="0"
              y="0"
            />
            <rect
              class="highcharts-plot-background"
              fill="none"
              height="250"
              width="580"
              x="10"
              y="50"
            />
            <rect
              class="highcharts-plot-border"
              data-z-index="1"
              fill="none"
              height="250"
              stroke="#cccccc"
              stroke-width="0"
              width="580"
              x="10"
              y="50"
            />
            <g
              class="highcharts-series-group"
              data-z-index="3"
            />
            <text
              class="highcharts-title"
              data-z-index="4"
              style="color: rgb(51, 51, 51); font-size: 18px; fill: #333333;"
              text-anchor="middle"
              x="300"
              y="64"
            >
              都道府県別 総人口数
            </text>
            <text
              class="highcharts-subtitle"
              data-z-index="4"
              style="color: rgb(102, 102, 102); fill: #666666;"
              text-anchor="middle"
              x="300"
              y="50"
            />
            <text
              class="highcharts-caption"
              data-z-index="4"
              style="color: rgb(102, 102, 102); fill: #666666;"
              text-anchor="start"
              x="10"
              y="312"
            />
            <g
              class="highcharts-legend highcharts-no-tooltip"
              data-z-index="7"
              visibility="hidden"
            >
              <rect
                class="highcharts-legend-box"
                fill="none"
                height="8"
                rx="0"
                ry="0"
                stroke="#999999"
                stroke-width="0"
                width="8"
                x="0"
                y="0"
              />
              <g
                data-z-index="1"
              >
                <g />
              </g>
            </g>
            <text
              class="highcharts-credits"
              data-z-index="8"
              style="cursor: pointer; color: rgb(153, 153, 153); font-size: 9px; fill: #999999;"
              text-anchor="end"
              x="590"
              y="395"
            >
              Highcharts.com
            </text>
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Mainのテスト 都道府県を選択(グラフ描画) 2`] = `
<DocumentFragment>
  <div
    class="title"
  >
    都道府県別 総人口
  </div>
  <div
    class="wrapper"
  >
    <div
      class="prefectures_area"
    >
      <div
        class="section_title"
      >
        都道府県選択
      </div>
      <div>
        <label>
          <input
            type="checkbox"
            value="1"
          />
          北海道
        </label>
      </div>
      <div>
        <label>
          <input
            type="checkbox"
            value="2"
          />
          青森県
        </label>
      </div>
      <div>
        <label>
          <input
            type="checkbox"
            value="3"
          />
          岩手県
        </label>
      </div>
    </div>
    <div
      class="graph_area"
    >
      <div
        data-highcharts-chart="1"
        style="height: 100%; overflow: hidden;"
      >
        <div
          class="highcharts-container "
          dir="ltr"
          id="highcharts-2"
          style="position: relative; overflow: hidden; width: 600px; height: 400px; text-align: left; line-height: normal; z-index: 0; user-select: none; outline: none;"
        >
          <svg
            aria-label="都道府県別 総人口数"
            class="highcharts-root"
            height="400"
            role="img"
            style="font-family: \\"Lucida Grande\\", \\"Lucida Sans Unicode\\", Arial, Helvetica, sans-serif; font-size: 12px;"
            version="1.1"
            viewBox="0 0 600 400"
            width="600"
            xmlns="http://www.w3.org/2000/svg"
          >
            <desc>
              Created with Highcharts 10.3.3
            </desc>
            <defs>
              <clippath
                id="highcharts-3-"
              >
                <rect
                  fill="none"
                  height="NaN"
                  width="NaN"
                  x="0"
                  y="0"
                />
              </clippath>
              <clippath
                id="highcharts-7-"
              >
                <rect
                  fill="none"
                  height="NaN"
                  width="NaN"
                  x="0"
                  y="0"
                />
              </clippath>
              <clippath
                id="highcharts-8-"
              >
                <rect
                  class="highcharts-animating"
                  fill="none"
                  height="NaNNaN"
                  width="NaN"
                  x="0"
                  y="0"
                />
              </clippath>
              <clippath
                id="highcharts-9-"
              >
                <rect
                  fill="none"
                  height="NaN"
                  width="99"
                  x="-99"
                  y="-99"
                />
              </clippath>
            </defs>
            <rect
              class="highcharts-background"
              fill="#ffffff"
              height="400"
              rx="0"
              ry="0"
              width="600"
              x="0"
              y="0"
            />
            <rect
              class="highcharts-plot-background"
              fill="none"
              height="NaN"
              width="NaN"
              x="NaN"
              y="50"
            />
            <rect
              class="highcharts-plot-border"
              data-z-index="1"
              fill="none"
              height="250"
              stroke="#cccccc"
              stroke-width="0"
              width="580"
              x="10"
              y="50"
            />
            <g
              class="highcharts-grid highcharts-xaxis-grid"
              data-z-index="1"
            >
              <path
                class="highcharts-grid-line"
                d="M 0 0"
                data-z-index="1"
                fill="none"
                opacity="1"
                stroke="#e6e6e6"
                stroke-dasharray="none"
                stroke-width="0"
              />
              <path
                class="highcharts-grid-line"
                d="M 0 0"
                data-z-index="1"
                fill="none"
                opacity="1"
                stroke="#e6e6e6"
                stroke-dasharray="none"
                stroke-width="0"
              />
              <path
                class="highcharts-grid-line"
                d="M 0 0"
                data-z-index="1"
                fill="none"
                opacity="1"
                stroke="#e6e6e6"
                stroke-dasharray="none"
                stroke-width="0"
              />
              <path
                class="highcharts-grid-line"
                d="M 0 0"
                data-z-index="1"
                fill="none"
                opacity="1"
                stroke="#e6e6e6"
                stroke-dasharray="none"
                stroke-width="0"
              />
              <path
                class="highcharts-grid-line"
                d="M 0 0"
                data-z-index="1"
                fill="none"
                opacity="1"
                stroke="#e6e6e6"
                stroke-dasharray="none"
                stroke-width="0"
              />
              <path
                class="highcharts-grid-line"
                d="M 0 0"
                data-z-index="1"
                fill="none"
                opacity="1"
                stroke="#e6e6e6"
                stroke-dasharray="none"
                stroke-width="0"
              />
            </g>
            <g
              class="highcharts-grid highcharts-yaxis-grid"
              data-z-index="1"
            >
              <path
                class="highcharts-grid-line"
                d="M 0 0"
                data-z-index="1"
                fill="none"
                opacity="1"
                stroke="#e6e6e6"
                stroke-dasharray="none"
                stroke-width="1"
              />
              <path
                class="highcharts-grid-line"
                d="M 0 0"
                data-z-index="1"
                fill="none"
                opacity="1"
                stroke="#e6e6e6"
                stroke-dasharray="none"
                stroke-width="1"
              />
              <path
                class="highcharts-grid-line"
                d="M 0 0"
                data-z-index="1"
                fill="none"
                opacity="1"
                stroke="#e6e6e6"
                stroke-dasharray="none"
                stroke-width="1"
              />
              <path
                class="highcharts-grid-line"
                d="M 0 0"
                data-z-index="1"
                fill="none"
                opacity="1"
                stroke="#e6e6e6"
                stroke-dasharray="none"
                stroke-width="1"
              />
              <path
                class="highcharts-grid-line"
                d="M 0 0"
                data-z-index="1"
                fill="none"
                opacity="1"
                stroke="#e6e6e6"
                stroke-dasharray="none"
                stroke-width="1"
              />
              <path
                class="highcharts-grid-line"
                d="M 0 0"
                data-z-index="1"
                fill="none"
                opacity="1"
                stroke="#e6e6e6"
                stroke-dasharray="none"
                stroke-width="1"
              />
              <path
                class="highcharts-grid-line"
                d="M 0 0"
                data-z-index="1"
                fill="none"
                opacity="1"
                stroke="#e6e6e6"
                stroke-dasharray="none"
                stroke-width="1"
              />
            </g>
            <g
              class="highcharts-axis highcharts-xaxis"
              data-z-index="2"
            >
              <path
                class="highcharts-tick"
                d="M 0 0"
                fill="none"
                opacity="1"
                stroke="#ccd6eb"
                stroke-width="1"
              />
              <path
                class="highcharts-tick"
                d="M 0 0"
                fill="none"
                opacity="1"
                stroke="#ccd6eb"
                stroke-width="1"
              />
              <path
                class="highcharts-tick"
                d="M 0 0"
                fill="none"
                opacity="1"
                stroke="#ccd6eb"
                stroke-width="1"
              />
              <path
                class="highcharts-tick"
                d="M 0 0"
                fill="none"
                opacity="1"
                stroke="#ccd6eb"
                stroke-width="1"
              />
              <path
                class="highcharts-tick"
                d="M 0 0"
                fill="none"
                opacity="1"
                stroke="#ccd6eb"
                stroke-width="1"
              />
              <path
                class="highcharts-tick"
                d="M 0 0"
                fill="none"
                opacity="1"
                stroke="#ccd6eb"
                stroke-width="1"
              />
              <text
                class="highcharts-axis-title"
                data-z-index="7"
                style="color: rgb(102, 102, 102); fill: #666666;"
                text-anchor="end"
                transform="translate(0,0)"
                x="NaN"
                y="NaN"
              >
                年度
              </text>
              <path
                class="highcharts-axis-line"
                d="M 0 0"
                data-z-index="7"
                fill="none"
                stroke="#ccd6eb"
                stroke-width="1"
              />
            </g>
            <g
              class="highcharts-axis highcharts-yaxis"
              data-z-index="2"
            >
              <text
                class="highcharts-axis-title"
                data-z-index="7"
                style="color: rgb(102, 102, 102); fill: #666666;"
                text-anchor="end"
                transform="translate(0,0)"
                x="NaN"
                y="30"
              >
                総人口数
              </text>
              <path
                class="highcharts-axis-line"
                d="M 0 0"
                data-z-index="7"
                fill="none"
                stroke="#ccd6eb"
                stroke-width="0"
              />
            </g>
            <g
              class="highcharts-series-group"
              data-z-index="3"
            >
              <g
                class="highcharts-series highcharts-series-0 highcharts-line-series highcharts-color-0"
                clip-path="url(#highcharts-8-)"
                data-z-index="0.1"
                opacity="1"
                transform="translate(NaN,50) scale(1 1)"
              >
                <path
                  class="highcharts-graph"
                  d="M -100000 -100000 L -100000 -100000 L -100000 -100000"
                  data-z-index="1"
                  fill="none"
                  stroke="#7cb5ec"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
                <path
                  class="highcharts-tracker-line"
                  d="M -100000 -100000 L -100000 -100000 L -100000 -100000"
                  data-z-index="2"
                  fill="none"
                  stroke="rgba(192,192,192,0.002)"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="22"
                />
              </g>
              <g
                class="highcharts-markers highcharts-series-0 highcharts-line-series highcharts-color-0 highcharts-tracker"
                clip-path="url(#highcharts-9-)"
                data-z-index="0.1"
                opacity="1"
                transform="translate(NaN,50) scale(1 1)"
              />
            </g>
            <text
              class="highcharts-title"
              data-z-index="4"
              style="color: rgb(51, 51, 51); font-size: 18px; fill: #333333;"
              text-anchor="middle"
              x="300"
              y="64"
            >
              都道府県別 総人口数
            </text>
            <text
              class="highcharts-subtitle"
              data-z-index="4"
              style="color: rgb(102, 102, 102); fill: #666666;"
              text-anchor="middle"
              x="300"
              y="50"
            />
            <text
              class="highcharts-caption"
              data-z-index="4"
              style="color: rgb(102, 102, 102); fill: #666666;"
              text-anchor="start"
              x="10"
              y="312"
            />
            <g
              class="highcharts-legend highcharts-no-tooltip"
              data-z-index="7"
              transform="translate(590,162)"
            >
              <rect
                class="highcharts-legend-box"
                fill="none"
                height="8"
                rx="0"
                ry="0"
                stroke="#999999"
                stroke-width="0"
                width="8"
                x="0"
                y="0"
              />
              <g
                data-z-index="1"
              >
                <g>
                  <g
                    class="highcharts-legend-item highcharts-line-series highcharts-color-0 highcharts-series-0"
                    data-z-index="1"
                    transform="translate(8,3)"
                  >
                    <path
                      class="highcharts-graph"
                      d="M 1 11 L 15 11"
                      fill="none"
                      stroke="#7cb5ec"
                      stroke-linecap="round"
                      stroke-width="2"
                    />
                    <path
                      class="highcharts-point"
                      d="M 8 15 A 4 4 0 1 1 8.003999999333336 14.999998000000167 Z"
                      fill="#7cb5ec"
                      opacity="1"
                      stroke="#ffffff"
                      stroke-width="0"
                    />
                    <text
                      data-z-index="2"
                      style="color: rgb(51, 51, 51); cursor: pointer; font-size: 12px; font-weight: bold; fill: #333333;"
                      text-anchor="start"
                      x="21"
                      y="15"
                    >
                      北海道
                    </text>
                  </g>
                </g>
              </g>
            </g>
            <g
              class="highcharts-axis-labels highcharts-xaxis-labels"
              data-z-index="7"
            >
              <text
                style="color: rgb(102, 102, 102); cursor: default; font-size: 11px; fill: #666666;"
                text-anchor="middle"
                transform="translate(0,0)"
                x="0"
              >
                1980
              </text>
              <text
                style="color: rgb(102, 102, 102); cursor: default; font-size: 11px; fill: #666666;"
                text-anchor="middle"
                transform="translate(0,0)"
                x="0"
              >
                1982
              </text>
              <text
                style="color: rgb(102, 102, 102); cursor: default; font-size: 11px; fill: #666666;"
                text-anchor="middle"
                transform="translate(0,0)"
                x="0"
              >
                1984
              </text>
              <text
                style="color: rgb(102, 102, 102); cursor: default; font-size: 11px; fill: #666666;"
                text-anchor="middle"
                transform="translate(0,0)"
                x="0"
              >
                1986
              </text>
              <text
                style="color: rgb(102, 102, 102); cursor: default; font-size: 11px; fill: #666666;"
                text-anchor="middle"
                transform="translate(0,0)"
                x="0"
              >
                1988
              </text>
              <text
                style="color: rgb(102, 102, 102); cursor: default; font-size: 11px; fill: #666666;"
                text-anchor="middle"
                transform="translate(0,0)"
                x="0"
              >
                1990
              </text>
            </g>
            <g
              class="highcharts-axis-labels highcharts-yaxis-labels"
              data-z-index="7"
            >
              <text
                style="color: rgb(102, 102, 102); cursor: default; font-size: 11px; fill: #666666;"
                text-anchor="end"
                transform="translate(0,0)"
                x="0"
              >
                500
              </text>
              <text
                style="color: rgb(102, 102, 102); cursor: default; font-size: 11px; fill: #666666;"
                text-anchor="end"
                transform="translate(0,0)"
                x="0"
              >
                1000
              </text>
              <text
                style="color: rgb(102, 102, 102); cursor: default; font-size: 11px; fill: #666666;"
                text-anchor="end"
                transform="translate(0,0)"
                x="0"
              >
                1500
              </text>
              <text
                style="color: rgb(102, 102, 102); cursor: default; font-size: 11px; fill: #666666;"
                text-anchor="end"
                transform="translate(0,0)"
                x="0"
              >
                2000
              </text>
              <text
                style="color: rgb(102, 102, 102); cursor: default; font-size: 11px; fill: #666666;"
                text-anchor="end"
                transform="translate(0,0)"
                x="0"
              >
                2500
              </text>
              <text
                style="color: rgb(102, 102, 102); cursor: default; font-size: 11px; fill: #666666;"
                text-anchor="end"
                transform="translate(0,0)"
                x="0"
              >
                3000
              </text>
              <text
                style="color: rgb(102, 102, 102); cursor: default; font-size: 11px; fill: #666666;"
                text-anchor="end"
                transform="translate(0,0)"
                x="0"
              >
                3500
              </text>
            </g>
            <text
              class="highcharts-credits"
              data-z-index="8"
              style="cursor: pointer; color: rgb(153, 153, 153); font-size: 9px; fill: #999999;"
              text-anchor="end"
              x="590"
              y="395"
            >
              Highcharts.com
            </text>
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
